state/etcd/etcd.go

```diff
@@ -0,0 +1,481 @@
+// Etcd is a state store implementation for Etcd.
+type Etcd struct {
+	state.DefaultBulkStore
+	client        *clientv3.Client
+	keyPrefixPath string
+	features      []state.Feature
+	logger        logger.Logger
+}
+
+type etcdConfig struct {
+	Endpoints     string `json:"endpoints"`
+	KeyPrefixPath string `json:"keyPrefixPath"`
+	// TLS
+	TlsEnable   string `json:"tlsEnable"`
+	Ca          string `json:"ca"`
+	Cert        string `json:"cert"`
+	Key         string `json:"key"`
+	KeyPassword string `json:"keyPassword"`
+}
+
+// NewEtcdStateStore returns a new etcd state store.
+func NewEtcdStateStore(logger logger.Logger) state.Store {
+	s := &Etcd{
+		logger:   logger,
+		features: []state.Feature{state.FeatureETag, state.FeatureTransactional},
+	}
+	s.DefaultBulkStore = state.NewDefaultBulkStore(s)
+
+	return s
+}
+
+// Init does metadata and config parsing and initializes the
+// Etcd client.
+func (e *Etcd) Init(metadata state.Metadata) error {
+	etcdConfig, err := metadataToConfig(metadata.Properties)
+	if err != nil {
+		return fmt.Errorf("couldn't convert metadata properties: %s", err)
+	}
+
+	e.client, err = e.ParseClientFromConfig(etcdConfig)
+	if err != nil {
+		return fmt.Errorf("initializing etcd client: %s", err)
+	}
+
+	e.keyPrefixPath = etcdConfig.KeyPrefixPath
+
+	return nil
+}
+
+func (e *Etcd) ParseClientFromConfig(etcdConfig *etcdConfig) (*clientv3.Client, error) {
+	endpoints := strings.Split(etcdConfig.Endpoints, ",")
+	if len(endpoints) == 0 || endpoints[0] == "" {
+		return nil, fmt.Errorf("endpoints required")
+	}
+
+	var tlsConfig *tls.Config
+	if etcdConfig.TlsEnable == tlsEnable {
+		if etcdConfig.Cert != "" && etcdConfig.Key != "" && etcdConfig.Ca != "" {
+			var err error
+			tlsConfig, err = utils.NewTLSConfigWithPassword(etcdConfig.Cert, etcdConfig.Key, etcdConfig.KeyPassword, etcdConfig.Ca)
+			if err != nil {
+				return nil, fmt.Errorf("tls authentication error: %s", err)
+			}
+		} else {
+			return nil, fmt.Errorf("tls authentication information is incomplete")
+		}
+	}
+
+	config := clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: 5 * time.Second,
+		TLS:         tlsConfig,
+	}
+	client, err := clientv3.New(config)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+// Features returns the features available in this state store.
+func (e *Etcd) Features() []state.Feature {
+	return e.features
+}
+
+func metadataToConfig(connInfo map[string]string) (*etcdConfig, error) {
+	m := &etcdConfig{}
+	err := metadata.DecodeMetadata(connInfo, m)
+	return m, err
+}
+
+// Get retrieves a Etcd KV item.
+func (e *Etcd) Get(ctx context.Context, req *state.GetRequest) (*state.GetResponse, error) {
+	keyWithPath := e.keyPrefixPath + "/" + req.Key
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	resp, err := e.client.Get(ctx, keyWithPath)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get key %s: %s", keyWithPath, err)
+	}
+
+	if resp == nil || resp.Kvs == nil {
```

Comment:
```suggestion
	if resp == nil || len(resp.Kvs) == 0 {
```

This way if `resp.Kvs` is non-nil, but an empty array, you won't panic below